receiver/nats.go

```diff
@@ -0,0 +1,159 @@
+features in the nats golang client. Basic queue groups is also supported.
+*/
+type Nats struct {
+	Handler       skogul.HandlerRef `doc:"Handler used to parse, transform and send data."`
+	Servers       string            `doc:"Comma separated list of nats URLs"`
+	Subject       string            `doc:"Subject to subscribe to messages on"`
+	Queue         string            `doc:"Worker queue to distribute messages on"`
+	Name          string            `doc:"Client name"`
+	Username      string            `doc:"Client username"`
+	Password      string            `doc:"Client password"`
+	TLSClientKey  string            `doc:"TLS client key file path"`
+	TLSClientCert string            `doc:"TLS client cert file path"`
+	TLSCACert     string            `doc:"CA cert file path"`
+	UserCreds     string            `doc:"Nats credentials file path"`
+	NKeyFile      string            `doc:"Nats nkey file path"`
+	Insecure      bool              `doc:"TLS InsecureSkipVerify"`
+	conOpts       *[]nats.Option
+	natsCon       *nats.Conn
+	wg            sync.WaitGroup
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+	if n.Handler.Name == "" {
+		return skogul.MissingArgument("Handler")
+	}
+	if n.Subject == "" {
+		return skogul.MissingArgument("Subject")
+	}
+	if n.Servers == "" {
+		return skogul.MissingArgument("Servers")
+	}
+	//User Credentials
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		//Cred file contains nkey.
+		return fmt.Errorf("Please configure usercreds or nkeyfile.")
+	}
+
+	return nil
+}
+
+func (n *Nats) Start() error {
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.conOpts = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.UserCreds != "" {
+		*n.conOpts = append(*n.conOpts, nats.UserCredentials(n.UserCreds))
+	}
+
+	//Plain text passwords
+	if n.Username != "" && n.Password != "" {
+		if n.TLSClientKey != "" {
+			natsLog.Warnf("Using plain text password over a non encrypted transport!")
+		}
+		*n.conOpts = append(*n.conOpts, nats.UserInfo(n.Username, n.Password))
+	}
+	//TLS authentication, Note: Fix selfsigned certificates.
+	if n.TLSClientKey != "" && n.TLSClientCert != "" {
+		cert, err := tls.LoadX509KeyPair(n.TLSClientCert, n.TLSClientKey)
+		if err != nil {
+			return fmt.Errorf("error parsing X509 certificate/key pair: %v", err)
+		}
+
+		cp, err := skogul.GetCertPool(n.TLSCACert)
+		if err != nil {
+			return fmt.Errorf("Failed to initialize root CA pool")
+		}
+
+		config := &tls.Config{
+			InsecureSkipVerify: n.Insecure,
+			Certificates:       []tls.Certificate{cert},
+			RootCAs:            cp,
+		}
+		*n.conOpts = append(*n.conOpts, nats.Secure(config))
+	}
+
+	//NKey auth
+	if n.NKeyFile != "" {
+		opt, err := nats.NkeyOptionFromSeed(n.NKeyFile)
+		if err != nil {
+			natsLog.Fatal(err)
+		}
+		*n.conOpts = append(*n.conOpts, opt)
+	}
+
+	//Log disconnects
+	*n.conOpts = append(*n.conOpts, nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
+		natsLog.WithError(err).Error("Got disconnected!")
+	}))
```

Comment:
Is it always an error if a client disconnects? I'm thinking of controlled restarts of the client or similar, e.g. maintenance/patching. Or will that not trigger a "disconnect" but rather a "bye" of some sort?

I'm thinking that if this logs an error when a client "cleanly disconnects", it's noisy. But if this doesn't trigger on a "clean disconnect", I'm fine with it.

Reply:
Good point, i had not verified it, but the API has an "DisconnectHandler" which seem to be invoked on every disconnect, rather than if an error occurred as in the case of DisconnectErrHandler. 

Did a quick local test as well, which confirmed that it wont trigger on a "clean" disconnect.